src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java

```diff
@@ -166,10 +171,79 @@ private boolean validateNucleusConfig(CompletableFuture<DeploymentResult> totall
                         .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE, e));
                 return false;
             }
+
+            Integer timeoutSec = deployment.getDeploymentDocumentObj().getConfigurationValidationPolicy().timeoutInSeconds();
+            long configTimeout = Duration.ofSeconds(20).toMillis();
+            if (timeoutSec != null) {
+                configTimeout = Duration.ofSeconds(timeoutSec).toMillis();
+            }
+
+            Map<String, Object> currentConfig = kernel.getConfig().toPOJO();
+            if (configTimeout == 0
+                    || !deviceConfiguration.isDeviceConfiguredToTalkToCloud()
+                    || !hasNucleusConfigurationChange(currentConfig, nucleusConfig)) {
+                logger.atInfo().log("Skipping connectivity validation");
+                return true;
+            }
+            try {
+                // Update Nucleus with new config
+                logger.atInfo().log("Applying Nucleus config");
+                Map<String, Object> modifiedConfig = kernel.getConfig().toPOJO();
+                replaceNucleusConfiguration(modifiedConfig, nucleusConfig);
+                activator.mergeConfig(modifiedConfig, deployment);
+                TimeUnit.SECONDS.sleep(5);
```

Comment:
seems like you should use an `executerService` to execute the config merge and then wait for that task to finish before continuing your validation.
Also, I think you should be able to roll back to known configuration if deployment is cancelled during this wait time.